### PR TITLE
Add basic structure

### PR DESCRIPTION
**Pull request by SimonNumberTwo**
# Description
This PR adds a basic structure to the framework. It adds test cases to ensure the correct behaviour. Please keep in mind that this is still work in progress and will change over time but please see it as a first version we can then continue to work with.

### CODE REVIEW

File: Sources/Jetworking/Client.swift [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        guard let urlResponse = urlResponse as? HTTPURLResponse else { return completion(.failure(APIError.responseMissing)) }
+
+        let statusCode = HTTPStatusCodeType(statusCode: urlResponse.statusCode)
+
+        switch statusCode {
+        case .successful:
+            guard
+                let data = data,
+                let decodedData = try? clientConfiguration.decoder.decode(ResponseType.self, from: data)
+            else { return completion(.failure(APIError.decodingError)) }
```
Comment (JensK611) [inline Sources/Jetworking/Client.swift:138]: I suggest to pass the underlying error of the decoding error, then it is possible to look into it which property couldn't decoded. Maybe as parameter of the decodingError case. But then we have to use `try` without optional.

File: Sources/Jetworking/Client.swift [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            // Evaluate Header fields --> urlResponse.allHeaderFields
+
+            completion(.success(decodedData))
+            
+        case .clientError, .serverError:
+            guard let error = error else { return }
```
Comment (JensK611) [inline Sources/Jetworking/Client.swift:144]: The completion should always be called! if we don't know which error we received, we could introduce some `unexpectedError` or similar.

File: Sources/Jetworking/ClientConfiguration.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+public struct ClientConfiguration {
+    let baseURL: URL
+    let authenticationMethod: AuthenticationMethod
+    let encoder: JSONEncoder
```
Comment (JensK611) [inline Sources/Jetworking/ClientConfiguration.swift:6]: In the future we should consider if we just use Encoder and Decoder. The Api could return xml for example.

File: Sources/Jetworking/Globals/HTTPStatusCodeType.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
+enum HTTPStatusCodeType {
+    case informational
+    case successful
+    case redirection
+    case clientError
+    case serverError
+    case unknown
+
+    init(statusCode: Int) {
```
Comment (JensK611) [inline Sources/Jetworking/Globals/HTTPStatusCodeType.swift:9]: We should make it possible to access the statusCode in the Result as well! Sometimes it is necessary to differentiate between `Ok` and `Created` for example.

File: Sources/Jetworking/Client.swift [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        let request: URLRequest = createRequest(forHttpMethod: .POST, andPathComponent: endpoint.pathComponent)
+        let bodyData: Data = try! clientConfiguration.encoder.encode(body) // TODO: remove force unwrap --> Error handling
+        let dataTask = session.uploadTask(with: request, from: bodyData) { [weak self] data, urlResponse, error in
+            self?.handleResponse(data: data, urlResponse: urlResponse, error: error, endpoint: endpoint, completion: completion)
+        }
+
+        dataTask.resume()
+    }
+
+    public func put<BodyType: Codable, ResponseType>(endpoint: Endpoint<ResponseType>, body: BodyType, _ completion: @escaping (Result<ResponseType, Error>) -> Void) {
```
Comment (JensK611) [inline Sources/Jetworking/Client.swift:41]: BodyType should only constraint to `Encodable`

File: Sources/Jetworking/Client.swift [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        let request: URLRequest = createRequest(forHttpMethod: .PUT, andPathComponent: endpoint.pathComponent)
+        let bodyData: Data = try! clientConfiguration.encoder.encode(body)
```
Comment (JensK611) [inline Sources/Jetworking/Client.swift:43]: Add todo here as well!

File: Sources/Jetworking/Client.swift [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            url: clientConfiguration.baseURL.appendingPathComponent(pathComponent),
+            httpMethod: httpMethod,
+            headerFields: getHeaderFields()
+        )
+    }
+
+    private func getHeaderFields() -> [String: String] {
+        var headerFields: [String: String] = [
+            "Accept": "application/json",
+            "Content-Type": "application/json"
```
Comment (JensK611) [inline Sources/Jetworking/Client.swift:81]: We should make the default header fields exchangeable.
What do you think abound placing them in the ClientConfiguration together with a static default value which is returning this map? Then we could adjust them in the initializer if we need to?

File: Sources/Jetworking/Client.swift [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        switch clientConfiguration.authenticationMethod {
+        case .none:
+            return nil
+
+        case let .basicAuthentication(username, password):
+            var authString: String = ""
+            let credentialsString = "\(username):\(password)"
+            if let credentialsData = credentialsString.data(using: .utf8) {
+                let base64Credentials = credentialsData.base64EncodedString(options: [])
+                authString = "Basic \(base64Credentials)"
```
Comment (JensK611) [inline Sources/Jetworking/Client.swift:107]: This should be extracted and tested as well.

File: Sources/Jetworking/ClientConfiguration.swift (hunk 3 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/ClientConfiguration.swift:6]: Unfortunately this is not possible as `JSONEncoder` or `JSONDecoder` are not dependent on `Encoder` or `Decoder`. Furthermore when using `Codable` objects are not usable with xml responses so this has to be dealt with in a special way. For now I would like to stick with the most common use case of having JSON data to send and retrieve.

File: Sources/Jetworking/Client.swift (hunk 5 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/Client.swift:41]: Done.

File: Sources/Jetworking/Client.swift (hunk 2 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/Client.swift:144]: Done.

File: Sources/Jetworking/Client.swift (hunk 6 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/Client.swift:43]: Done.

File: Sources/Jetworking/Client.swift (hunk 7 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/Client.swift:81]: In my opinion we should tackle this in a future ticket. This MR just adds a first version of the framework. I will create tickets later on where we can discuss the different tasks. Maybe we only need some header fields in some cases and not always etc.

File: Sources/Jetworking/Client.swift (hunk 8 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/Client.swift:107]: I will add a ticket to add tests, for now I would leave it as is if that's okay for you.

File: Sources/Jetworking/Globals/HTTPStatusCodeType.swift (hunk 4 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/Globals/HTTPStatusCodeType.swift:9]: I will add a ticket for this.

File: Sources/Jetworking/Client.swift (hunk 1 above)
Comment (SimonNumberTwo) [reply Sources/Jetworking/Client.swift:138]: This is just a first version of the framework. I will add tickets so we create a proper error handling. Maybe this should also be discussed in a larger group of contributors.

File: Sources/Jetworking/ClientConfiguration.swift (hunk 3 above)
Comment (JensK611) [reply Sources/Jetworking/ClientConfiguration.swift:6]: Yep that's totally fine!

File: Sources/Jetworking/Globals/HTTPStatusCodeType.swift (hunk 4 above)
Comment (JensK611) [reply Sources/Jetworking/Globals/HTTPStatusCodeType.swift:9]: Ok perfect!

File: Sources/Jetworking/Client.swift (hunk 8 above)
Comment (JensK611) [reply Sources/Jetworking/Client.swift:108]: Yep that's ok!

**Pull request merged**